scripts/install-build-tools.sh

```diff
@@ -6,19 +6,74 @@ green="\033[0;32m"
 cyan="\033[0;36m"
 end="\033[0m"
 
-set -e
+# for debugging, 'set -x' can be added to trace the commands
+set -ex
 
 SUDO=''
 if (( $EUID != 0 )); then
     echo -e "non-root user, sudo required"
     SUDO='sudo'
 fi
 
+# as of June 2024, python3.13 does not support pylint so limiting most recent version 3.12
+PYTHON_VERSIONS=("3.12" "3.11" "3.10")
+
+# Function to install Python and necessary packages
+install_python() {
+  local PY_VERS=$1
+  $SUDO apt install -y python${PY_VERS} python${PY_VERS}-distutils
+  wget https://bootstrap.pypa.io/get-pip.py
+  python${PY_VERS} get-pip.py
+  rm get-pip.py
+  python${PY_VERS} -m pip install pylint matplotlib
+}
+
+echo "Python3 at start of script version: $(python3 --version)"
+echo "OS Type: $OSTYPE"
+DEFAULT_PYTHON=`python3 --version | grep -E 'Python 3.[0-9]+'`
+
+
 if [[ "$OSTYPE" == "darwin"* ]]; then
   CPUS=$(sysctl -n hw.ncpu)
   # ensure development environment is set correctly for clang
   $SUDO xcode-select -switch /Library/Developer/CommandLineTools
-  brew install llvm@14 googletest google-benchmark lcov make wget cmake curl
+
+  echo "Running sudo with homebrew as root through a shell script is not allowed."
+  echo "These installs can be done manually in a terminal, however."
```

Comment:
Just to clarify for me: does this mean that macOS users need to run these commands manually, outside the shell script, as an additional setup step? Is that documented somewhere else (I see a note mentioning that running `brew` as root is not supported on macOS, but it doesn't clarify what the user needs to do)?

Reply:
So this is the error I encounter after I run the following and enter my password correctly in the mac terminal as a root user.
```console
$ sudo ./scripts/install-build-tools.sh 
$ Password:
Setting up build environment...
Python3 at start of script version: Python 3.13.0a2
OS Type: darwin23
Running sudo with homebrew as root through a shell script is not allowed.
These installs can be done manually in a terminal, however.
Error: Running Homebrew as root is extremely dangerous and no longer supported.
As Homebrew does not drop privileges on installation you would be giving all
build scripts full access to your system.
```
Should a README or alternative file be created for commands to run for macOS users where homebrew blocks standard installation through `sudo .scripts/install-build-tools.sh`?

Reply:
Ah, this is just saying that `install-build-tools.sh` needs to be run *without* `sudo` on macOS?

Reply:
Yes, that's what worked for me. Also the user will have to manually set their python version to 3.10-3.12 if not already done. The script will install one of those versions if not already installed and if found in the brew taps. We discussed it to be invasive to change .bashrc/.zshrc files and aliases, so I have this case:
```console
    elif [[ $(echo "${VERSION_SHORT} >= 3.13" | bc) ]]; then
      echo "Your Python3 default version: '$DEFAULT_PYTHON' is not supported by pylint yet"
      echo "Please set your default python3 to the recently installed version '${PY_INSTALLED}' and rerun this script."
      exit 1
    fi
```
Should also do the same comparison for versions <= 3.9.

Reply:
Note that that test either needs to be updated to ensure `0` and `1` are treated as truth values, or use `((`/`))` instead of `[[`/`]]`.